File: examples/cpp/memento.cpp
#include <iostream>
#include <memory>
#include <stack>
#include <string>

class EditorMemento {
    std::string text_;
public:
    EditorMemento(const std::string& text) : text_(text) {}
    const std::string& getText() const { return text_; }
};

class TextEditor {
    std::string text_;
public:
    void type(const std::string& words) { text_ += words; }
    std::string getText() const { return text_; }
    std::unique_ptr<EditorMemento> save() const {
        return std::make_unique<EditorMemento>(text_);
    }
    void restore(const EditorMemento& memento) {
        text_ = memento.getText();
    }
};

class History {
    std::stack<std::unique_ptr<EditorMemento>> snapshots_;
public:
    void push(std::unique_ptr<EditorMemento> memento) {
        snapshots_.push(std::move(memento));
    }
    std::unique_ptr<EditorMemento> pop() {
        if (snapshots_.empty()) return nullptr;
        auto m = std::move(snapshots_.top());
        snapshots_.pop();
        return m;
    }
};

int main() {
    std::cout << "=== Memento: Text Editor Undo ===" << std::endl;

    TextEditor editor;
    History history;

    editor.type("Hello");
    history.push(editor.save());
    std::cout << "Typed: \"" << editor.getText() << "\"" << std::endl;

    editor.type(", World");
    history.push(editor.save());
    std::cout << "Typed: \"" << editor.getText() << "\"" << std::endl;

    editor.type("!!!");
    std::cout << "Typed: \"" << editor.getText() << "\"" << std::endl;

    auto m = history.pop();
    if (m) editor.restore(*m);
    std::cout << "Undo:  \"" << editor.getText() << "\"" << std::endl;

    m = history.pop();
    if (m) editor.restore(*m);
    std::cout << "Undo:  \"" << editor.getText() << "\"" << std::endl;

    return 0;
}

File: examples/csharp/Patterns/Memento.cs
namespace DesignPatterns.Memento
{
    public class EditorMemento
    {
        public string Content { get; }
        public int CursorPosition { get; }
        public DateTime Timestamp { get; }

        public EditorMemento(string content, int cursorPos)
        {
            Content = content;
            CursorPosition = cursorPos;
            Timestamp = DateTime.Now;
        }
    }

    public class TextEditor
    {
        public string Content { get; set; } = "";
        public int CursorPosition { get; set; }

        public EditorMemento Save() => new(Content, CursorPosition);

        public void Restore(EditorMemento memento)
        {
            Content = memento.Content;
            CursorPosition = memento.CursorPosition;
        }

        public override string ToString() =>
            $"Editor(content=\"{Content}\", cursor={CursorPosition})";
    }

    public class History
    {
        private readonly Stack<EditorMemento> _snapshots = new();

        public void Push(EditorMemento memento) => _snapshots.Push(memento);

        public EditorMemento? Pop() =>
            _snapshots.Count > 0 ? _snapshots.Pop() : null;
    }

    public static class Demo
    {
        public static void Run()
        {
            var editor = new TextEditor();
            var history = new History();

            editor.Content = "Hello";
            editor.CursorPosition = 5;
            history.Push(editor.Save());
            Console.WriteLine($"  State 1: {editor}");

            editor.Content = "Hello World";
            editor.CursorPosition = 11;
            history.Push(editor.Save());
            Console.WriteLine($"  State 2: {editor}");

            editor.Content = "Hello World!!!";
            editor.CursorPosition = 14;
            Console.WriteLine($"  State 3: {editor}");

            Console.WriteLine("  --- Undo ---");
            var memento = history.Pop();
            if (memento != null) editor.Restore(memento);
            Console.WriteLine($"  After undo: {editor}");

            memento = history.Pop();
            if (memento != null) editor.Restore(memento);
            Console.WriteLine($"  After undo: {editor}");
        }
    }
}
